Add qiyam-ul-layl to prayer times

basememara opened the pull request:
Added dynamic qiyam-ul-layl time. Using last third of the night between maghrib and fajr for calculation.

For increased accuracy, should re-calculate prayers for next day to retrieve precise next day fajr, but traded off for performance and code simplicity since precision is fractional difference. However, could be a problem if next day requires daylight saving change or leap day.

Open for suggestions and discussion.

basememara committed:
Add qiyam-ul-layl to prayer times

basememara committed:
Minor code formatting fix

z3bi commented:
JazakAllah for the pull request. I went to double check that the time for Qiyam is the last third of the night and found that Qiyam (being a sunnah) doesn't have a specific time, just that it should take place shortly before Fajr. The last third of the night seems to come from a hadith that Allah descends to the lower heavens during the last third of the night, but there is also another hadith that mentions "The closest that a slave comes to his Lord is during the middle of the latter portion of the night".

I think it is useful to be able to know when the last third of the night is, the question comes in what the name should be. Plus it isn't in the Prayer enum (adding it there introduces other issues because again, the time for it isn't actually specific) so maybe it shouldn't be called Qiyam. 

Also I would say that the situation for DST changing between one day and another is significant enough that we should calculate the actual prayer times for the next day. To do this I think we would need to make date and calculationParameters actual members of the prayer time struct so that they can be accessed for creating tomorrow's prayer times. 

The fact that it is a computed property worries me though because someone may repeatedly access it, not understanding the computational expense incurred each time. Another way to go about this might be to separate it out into it's own struct and initialize it with a PrayerTime value. The act of initializing a new value might better convey the amount of computation going into it.

basememara committed:
Add sunnah times struct that accepts a prayer times instance.

basememara commented:
جزاك اللهُ خيرًا, I love the suggestions and perspective! I've updated the PR, pls give it some thought and let me know what you think.

basememara committed:
Updated sunnah times parameter label for initializer

z3bi commented on Swift/Adhan/Adhan.swift:
```diff
@@ -172,7 +172,7 @@ public enum CalculationMethod {
 
 /* Prayer times for a location and date using the given calculation parameters.
-All prayer times are in UTC and should be display using an NSDateFormatter that
+All prayer times are in UTC and should be display using an DateFormatter that
```
"using a DateFormatter"

z3bi commented on Swift/Adhan/Adhan.swift:
```diff
@@ -403,6 +408,35 @@ public struct Qibla {
 }
 
+/* Sunnah times for a location and date using the given prayer times.
+All prayer times are in UTC and should be display using an DateFormatter that
```
"using a DateFormatter"

z3bi commented on Swift/Adhan/Adhan.swift:
```diff
@@ -403,6 +408,35 @@ public struct Qibla {
+All prayer times are in UTC and should be display using an DateFormatter that
+has the correct timezone set. */
+public struct SunnahTimes {
+    public let midnight: Date
```
maybe rename to "middleOfTheNight" as to be distinct from 12:00 AM

z3bi commented on Swift/Adhan/Adhan.swift:
```diff
@@ -403,6 +408,35 @@ public struct Qibla {
+has the correct timezone set. */
+public struct SunnahTimes {
+    public let midnight: Date
+    public let lastThird: Date
```
for clarity perhaps "lastThirdOfTheNight", I know it's a bit longer but I think it's better to be more clear.

z3bi commented on Swift/Adhan/Adhan.swift:
```diff
@@ -831,10 +865,20 @@ struct TimeComponents {
+extension Calendar {
+    
+    /// All calculations are done using a gregorian calendar with the UTC timezone
+    static let utcGregorian: Calendar = {
```
I would name this "gregorianUTC"

basememara replied:
Funny, I went back and forth on this and couldn't decide. Ok will change it.

z3bi commented on Swift/Adhan/Adhan.swift:
```diff
@@ -403,6 +408,35 @@ public struct Qibla {
+                return nil
+        }
+        
+        let minutes = cal.dateComponents([.minute], from: prayerTimes.maghrib, to: nextDayPrayers.fajr).minute!
```
It is more efficient, and technically more accurate to use timeInterval and do the calculation in seconds. We can then do roundedMinute() at the very end.

z3bi commented:
I think this is a much better approach. I had a few comments, mostly about naming but this is looking good. Before final merging I would like for the unit tests to be expanded and cover some high latitude locations that use different high latitude rules (especially the middle of the night rule). And we should also have some tests that cover DST going into effect.

basememara commented:
Cool, great points and about the unit tests too. Will work on it and update the PR insha'Allah.

sgtsquiggs reviewed (commented):
Are there edge cases we should test? High latitudes, etc. A single test doesn't seem like enough. Great PR overall, great addition to Adhan!

sgtsquiggs commented:
Ameir beat me to it :+1:

basememara committed:
Update static Gregorian UTC API name

basememara committed:
Update sunnah time property name. Minor comment fix.

basememara committed:
Use seconds for sunnah time calculations for higher precision

basememara committed:
Add more unit tests for sunnah times

basememara commented:
I've updated the PR and added more unit tests, but I still need a better way to verify that the unit test results are actually right. I've started to use this as a baseline: https://www.islamicfinder.org. Let me know if there's a better way to verify the unit test results or even if is the right approach to what to test.

basememara commented on Swift/Adhan/Adhan.swift:
```diff
@@ -403,6 +408,35 @@ public struct Qibla {
+                return nil
+        }
+        
+        let seconds = cal.dateComponents([.second], from: prayerTimes.maghrib, to: nextDayPrayers.fajr).second!
```
I'm not sure if this is what you meant, but seconds are indeed more precise. I just got confused because I don't think I needed to use `roundedMinutes()` anywhere.

z3bi replied:
I actually meant using Date's timeIntervalSinceDate function which gives you an NSTimeInterval value of the difference in seconds. By using date components it has to do expensive calendrical calculations, when using timeIntervalSinceDate it merely subtracts two doubles which is much more efficient.

basememara replied:
Ohh, of course! Much better..

z3bi commented:
I don't think this necessarily needs an authoritative source for the unit test values. We could simply create tests for these edge case scenarios and assert that they are equal to a hard coded value that we have independently verified by hand. So we could look at the Maghrib time of the first day, look at the Fair time of the next day, calculate the difference and then validate that the struct has the same value we got manually, maybe mention what the Maghrib and Fair times were in comments so upon visual inspection we know where those values came from.

basememara committed:
Use time interval difference instead of components for difference

basememara committed:
Use UTC time interval instead of date components for date calculation

z3bi commented on Swift/Adhan/Adhan.swift:
```diff
@@ -403,6 +408,35 @@ public struct Qibla {
+                return nil
+        }
+        
+        let seconds = nextDayPrayers.fajr.timeIntervalSince1970 - prayerTimes.maghrib.timeIntervalSince1970
```
let seconds = nextDayPrayers.fajr.timeIntervalSince(prayerTimes.maghrib)

basememara replied:
Masha'Allah you're a star 👍

z3bi commented on Swift/Adhan/Adhan.swift:
```diff
@@ -403,6 +408,35 @@ public struct Qibla {
+        let seconds = nextDayPrayers.fajr.timeIntervalSince1970 - prayerTimes.maghrib.timeIntervalSince1970
+        
+        let half = seconds / 2.0
+        self.middleOfTheNight = Date(timeIntervalSince1970: prayerTimes.maghrib.timeIntervalSince1970 + half)
```
self.middleOfTheNight = prayerTimes.maghrib.addingTimeInterval(seconds/2)

z3bi commented on Swift/Adhan/Adhan.swift:
```diff
@@ -403,6 +408,35 @@ public struct Qibla {
+        self.middleOfTheNight = Date(timeIntervalSince1970: prayerTimes.maghrib.timeIntervalSince1970 + half)
+        
+        let twoThirds = seconds * (2.0 / 3.0)
+        self.lastThirdOfTheNight = Date(timeIntervalSince1970: prayerTimes.maghrib.timeIntervalSince1970 + twoThirds)
```
self.lastThirdOfTheNight = prayerTimes.maghrib.addingTimeInterval(seconds * (2/3))

basememara committed:
Refactored time interval calculations for sunnah times

basememara committed:
Fixed and verified unit tests for sunnah times

basememara commented:
I updated the unit tests. I hope it covers what you had in mind, but please feel free to let me know.

BTW, a pretty useful site for adding durations to dates for this kind of testing:
https://www.timeanddate.com/date/timeduration.html
https://www.timeanddate.com/date/timeadd.html?type=add

z3bi commented:
It's a good set of base unit tests, I may go through and add some more but I think it's enough for merging the PR. JazakAllah for the contribution!
